### PR TITLE
Fixit CLI

### LINKED ISSUE #147: Fixit CLI
Launch a brand new `fixit-cli` using `argparse` (Will fix  #124) 

Currently, all the scripts in `fixit/cli` are individual, which is good as per my plan of new CLI.

Here's my plan of implementing this:
The CLI will look like the following:

`setup.py`

```python
entry_points={
        'console_scripts': [
            'fixit-cli = cli.main:launch_cli'
        ],
    },
```

`fixit/cli/main.py`

```python
from fixit.cli import apply_fix

def _parse_flags(argv: List[str]) -> argparse.Namespace:
    parser = argparse_flags.ArgumentParser(prog="fixit", description="Fixit CLI ")
    # Register sub-commands
    subparser = parser.add_subparsers(title='command')

    # Will add all individual scripts i.e.
    add_new_rule.register_subparser(subparser) # add_new_rule
    apply_fix.register_subparser(subparser) # apply_fix
    ...
    return parser.parse_args(argv[1:])

def main(args: argparse.Namespace) -> None:
    args.subparser_fn(args)

def launch_cli() -> None:
    app.run(main, flags_parser=_parse_flags)

if __name__ == "__main__":
    launch_cli()
```

`fixit/cli/apply_fix.py`

```python
def register_subparser(parsers: argparse._SubParsersAction) -> None:
    """Add subparser for `apply_fix` command."""
    apply_fix_parser = parsers.add_parser(
        "apply_fix",
        help=APPLY_FIX_HELP_TEXT,
        parents=[get_paths_parser(), ...]
    )

    apply_fix_parser.set_defaults(subparser_fn=_apply_fixes)


def _apply_fixes(args: argparse.Namespace):
    # to-do
```

By doing this we can use fixit as a console script, which is very convenient.

I am planning to do this in small and individual PRs, starting from `fixit/cli/main.py` then each sub-commands (`apply_fix, add_new_rule`), etc.
Other files will stay as it is(`utils.py, args.py`).

Let me know if you want to add anything or have any suggestions on this. So, I can start ASAP

*Note: we can change CLI name later on or suggest the best name that suits well for this*

@jimmylai

### PR DESCRIPTION
**Pull request by acharles7**
## Summary
* Fixes #147 #124
* CLI for fixit 
* Supports both new CLI and existing scripts

Installation: 
```bash
$ pip install fixit
```
Usage: 
```bash
$ fixit --help
$ fixit run_rules --help
$ fixit apply_fix --help
$ fixit insert_suppressions --help
$ fixit add_new_rule --help
```

This PR creates new CLI for fixit module

Please follow this sequence to review this PR:
```
1. setup.py
2. main.py
3. {run_rules.py, apply_fix.py, insert_suppressions.py, add_new_rule.py}
```
*Note: To support backward compatibility, I have changed existing files `run_rules.py, apply_fix.py etc.` for this cli without affecting previous scripts workflow with a new architecture. Will remove redundant code once this CLI published and more robust.*

### CODE REVIEW

File: fixit/cli/add_new_rule.py [hunk 1]
```diff
@@ -113,5 +113,37 @@ def main() -> None:
...
+
+
+def _main(args: argparse.Namespace) -> None:
+    file_path = args.path
+    rule_name = args.name if args.name else file_path.stem
+    create_rule_file(file_path, rule_name)
+
+
 if __name__ == "__main__":
     main()
```
Comment (jimmylai) [inline fixit/cli/add_new_rule.py:149]: We can simply move the parser initialization here and then use `register_subparser` to initialize it.
That way we can remove the existing `main()` which is pretty much duplicated now.

```
parser = argparse.ArgumentParser(
        description="Creates a skeleton of adding new rule file"
    )
```


The same approach also applies to other scripts.

File: fixit/cli/main.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    spec: ModuleSpec = importlib.util.spec_from_file_location(
+        "version", directory / "_version.py"
+    )
+
+    version: ModuleType = importlib.util.module_from_spec(spec)
+    # pyre-ignore Pyre doesn't know about importlib entirely.
+    spec.loader.exec_module(version)
+    # pyre-ignore Pyre has no way of knowing that this constant exists.
+    return version.FIXIT_VERSION
```
Comment (jimmylai) [inline fixit/cli/main.py:38]: Can we simply do `from fixit._version import FIXIT_VERSION`?

File: fixit/cli/main.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+import argparse
+import importlib.util
+from importlib.machinery import ModuleSpec
+from pathlib import Path
+from types import ModuleType
+from typing import List
+
+from absl import app
+from absl.flags import argparse_flags
```
Comment (jimmylai) [inline fixit/cli/main.py:21]: Why do we need `absl`?
Doesn't the builtin `argparse` already provide the same features (`ArgumentParser`  and `add_subparsers`)?

**Review by jimmylai — CHANGES_REQUESTED** [review]

File: fixit/cli/main.py (hunk 2 above)
Comment (acharles7) [reply fixit/cli/main.py:38]: sure.

File: fixit/cli/main.py (hunk 3 above)
Comment (acharles7) [reply fixit/cli/main.py:21]: I have created CLI using `absl` in the past. So, implemented this way.
Will remove `absl` module and use only `argparse`.

Comment (acharles7) [reply fixit/cli/main.py]: removed `absl` support

File: fixit/cli/add_new_rule.py (hunk 1 above)
Comment (acharles7) [reply fixit/cli/add_new_rule.py:149]: As per current implementation, `register_subparser(parsers: argparse._SubParsersAction)` takes `subparser` object and add arguments to it and as per your suggestion, `register_subparser(parser)` will take `ArgumentParser ()` object. 

Can you please help me with how you want to achieve this in order to support backward compatibility? and to remove existing `main()` which contains duplicate arguments.

Comment (jimmylai) [reply fixit/cli/add_new_rule.py:149]: Both `argparse._SubParsersAction` and `ArgumentParser` have `add_parser` method.
So you can define the method as
```
from typing import Union
def register_subparser(parsers: Union[argparse._SubParsersAction, argparse.ArgumentParser]) -> None:
```

Comment (acharles7) [reply fixit/cli/add_new_rule.py:149]: `ArgumentParser` has no `add_parser()` method.
This is the one way I can achieve and we can still remove the existing `main()` which is duplicate. But, this way there is duplicate arguments (`add_argument()`)

Let me know if you have any suggestions.

```py
def register_subparser(parser: Union[argparse._SubParsersAction, argparse.ArgumentParser]):

    if isinstance(parser, argparse._SubParsersAction):
        add_new_rule_parser = parser.add_parser(
            "add_new_rule",
            help="Creates a skeleton of adding new rule file.",
        )
        add_new_rule_parser.add_argument( "--path", ..)
        add_new_rule_parser.add_argument( "--name", ..)

        add_new_rule_parser.set_defaults(subparser_fn=_main)

    else:
        parser.add_argument( "--path", ..)
        parser.add_argument( "--name", ..)

        args = parser.parse_args()
        _main(args)
```

```py
def _main(args: argparse.Namespace) -> None:
    file_path = args.path
    rule_name = args.name if args.name else file_path.stem
    create_rule_file(file_path, rule_name)
```

Comment (jimmylai) [reply fixit/cli/add_new_rule.py:149]: Sorry, I read it wrong. This is my suggestion:
So the common parts are those `add_argument` statement and the `set_defaults`.
If we simply let `register_subparser` take an optional `_SubParsersAction`, and it calls `argparse.ArgumentParser` to create a parser if parsers is `None`. Then we can remove existing duplicated `main()`.
Each `if __name__ == "__main__":` can simply calls `parser = register_subparser(); args = parser.parse_args(argv); args.subparser_fn(args)`

Comment (acharles7) [reply fixit/cli/add_new_rule.py:149]: How about this one? 

```py
def add_rule_parser():
    parser = argparse.ArgumentParser(
        description="Creates a skeleton of adding new rule file",
    )

    parser.add_argument( "--path" ..)
    parser.add_argument( "--name" ..)

    return parser


def register_subparser(parser: argparse._SubParsersAction=None) -> None:
    """Add subparser for `add_new_rule` command."""
    if parser is None:
        rule_parser = add_rule_parser()
        _main(rule_parser.parse_args())

    else:
        sub_parser = parser.add_parser(
            "add_new_rule",
            description="Creates a skeleton of adding new rule file.",
            help="Creates a skeleton of adding new rule file.",
        )
        sub_parser.add_argument( "--path" ..)
        sub_parser.add_argument( "--name" ..)
        
        sub_parser.set_defaults(subparser_fn=_main)


def _main(args: argparse.Namespace) -> None:
    file_path = args.path
    rule_name = args.name if args.name else file_path.stem
    create_rule_file(file_path, rule_name)


if __name__ == "__main__":
    register_subparser()
```

Comment (jimmylai) [reply fixit/cli/add_new_rule.py:149]: The `add_argument` parts are still duplicated which can be refactored as a function to be reused.

**Comment by acharles7** [comment]
@jimmylai Can you please look into the pyre error? To avoid those errors I have to put `pyre-ignore` everywhere. Do you know any workaround to avoid it?

File: fixit/cli/run_rules.py [hunk 4]
```diff
@@ -87,33 +100,48 @@ def get_formatted_reports_for_path(
...
-            get_use_ignore_comments_parser(),
-            get_skip_ignore_byte_marker_parser(),
-            get_compact_parser(),
-            get_multiprocessing_parser(),
-        ],
-    )
-
+def _parser_arguments(
+    parser: Union[argparse._SubParsersAction, argparse.ArgumentParser],
+    sub_parser: bool = True,
```
Comment (jimmylai) [inline fixit/cli/run_rules.py:105]: `sub_parser` is not needed.
You can simply do to help Pyre understand the type and it's more readable.
```
if isinstance(parser, argparse._SubParsersAction):
    ...
else:
    ...
```

Comment (acharles7) [reply fixit/cli/run_rules.py:105]: I tried this 
```
if isinstance(parser, argparse._SubParsersAction):
    ...
else:
    ...
```
but parser is always of type `argparse.ArgumentParser `, that's why I use `sub_parser`

File: fixit/cli/run_rules.py [hunk 5]
```diff
@@ -87,33 +100,48 @@ def get_formatted_reports_for_path(
...
+        _parser_arguments(run_rules_parser, sub_parser=False)
+
+    else:
+        run_rules_parser = parser.add_parser(
+            "run_rules",
+            description=DESCRIPTION,
+            parents=PARENTS,
+            help="Run fixit rules against python code",
+        )
+        _parser_arguments(run_rules_parser)
```
Comment (jimmylai) [inline fixit/cli/run_rules.py:141]: Move the logic to register_subparser and rename `_parser_arguments` as `_add_arguments` can keep the `_add_arguments` focuses on adding arguments as well as removing the type check.

```suggestion
def register_subparser(parser: argparse._SubParsersAction = None) -> None:
    """Add parser or subparser for `run_rules` command."""
    if parser is None:
        run_rules_parser = argparse.ArgumentParser(
            description=DESCRIPTION,
            parents=PARENTS,
        )
        _parser_arguments(run_rules_parser, sub_parser=False)
        _main(run_rules_parser.parse_args())
    else:
        run_rules_parser = parser.add_parser(
            "run_rules",
            description=DESCRIPTION,
            parents=PARENTS,
            help="Run fixit rules against python code",
        )
        _parser_arguments(run_rules_parser)
        run_rules_parser.set_defaults(subparser_fn=_main)
```

**Comment by acharles7** [comment]
> @jimmylai Can you please look into the pyre error? To avoid those errors I have to put `pyre-ignore` everywhere. Do you know any workaround to avoid it?

any update?

File: fixit/cli/add_new_rule.py [hunk 6]
```diff
@@ -86,32 +87,48 @@ def create_rule_file(file_path: Path, rule_name: str) -> None:
...
-        "-n",
         "--name",
         type=str,
         default="",
         help="Name of the rule, defaults to `New`",
     )
 
-    args = parser.parse_args()
 
+def register_subparser(parser: argparse._SubParsersAction = None) -> None:
```
Comment (jimmylai) [inline fixit/cli/add_new_rule.py:108]: The type is wrong.
It should be `Optional[argparse._SubParsersAction]`.

File: fixit/cli/add_new_rule.py [hunk 7]
```diff
@@ -86,32 +87,48 @@ def create_rule_file(file_path: Path, rule_name: str) -> None:
     print(f"Successfully created {file_path.name} rule file at {file_path.parent}")
 
 
-def main() -> None:
-    parser = argparse.ArgumentParser(
-        description="Creates a skeleton of adding new rule file"
-    )
+def _add_arguments(
+    parser: Union[argparse._SubParsersAction, argparse.ArgumentParser]
+) -> None:
```
Comment (jimmylai) [inline fixit/cli/add_new_rule.py:92]: The passed parser is simply an `ArgumentParser`.
```suggestion
def _add_arguments(
    parser: argparse.ArgumentParser
) -> None:
```

**Review by jimmylai — CHANGES_REQUESTED** [review]
Please follow the suggestion to fix the type error on all `_add_arguments` and `register_subparser`.

File: fixit/cli/add_new_rule.py (hunk 7 above)
Comment (acharles7) [reply fixit/cli/add_new_rule.py]: Done, thank you

File: fixit/cli/add_new_rule.py (hunk 6 above)
Comment (acharles7) [reply fixit/cli/add_new_rule.py]: Done

File: fixit/cli/run_rules.py [hunk 8]
```diff
@@ -87,33 +100,38 @@ def get_formatted_reports_for_path(
...
-    )
-
+def _add_arguments(parser: argparse.ArgumentParser) -> None:
+    """All required arguments for `run_rules`"""
     parser.add_argument(
         "--cache-timeout",
         type=int,
         help="Timeout (seconds) for metadata cache fetching. Default is 2 seconds.",
         default=2,
     )
```
Comment (jimmylai) [inline fixit/cli/run_rules.py:110]: This is duplicated to `get_metadata_cache_parser`, can you change to use `get_metadata_cache_parser`?
https://github.com/Instagram/Fixit/blob/18365db024258817f7f30006886c14ce8f07bbd5/fixit/cli/args.py#L174-L182

**Review by jimmylai — APPROVED** [review]
Thanks for contributing.

Found one more thing can be improved.

File: fixit/cli/run_rules.py (hunk 8 above)
Comment (acharles7) [reply fixit/cli/run_rules.py:110]: ok

Comment (acharles7) [reply fixit/cli/run_rules.py]: Done

File: fixit/cli/run_rules.py [hunk 9]
```diff
@@ -163,9 +178,6 @@ def main(raw_args: Sequence[str]) -> int:
             + f"{time.time() - start_time :.2f} seconds."
         )
 
-    # Return with an exit code of 1 if there are any violations found.
-    return int(bool(formatted_reports))
```
Comment (jimmylai) [inline fixit/cli/run_rules.py:167]: `run_rule` command can be run in builds and the return code will be used to fail the builds.
Can you keep this behavior unchanged?
It should also be added to `main.py`. Each _main can return an int as an argument for `sys.exit`.

File: fixit/cli/add_new_rule.py [hunk 10]
```diff
@@ -86,32 +88,48 @@ def create_rule_file(file_path: Path, rule_name: str) -> None:
...
     file_path = args.path
     rule_name = args.name if args.name else file_path.stem
     create_rule_file(file_path, rule_name)
 
+    return 0
+
 
 if __name__ == "__main__":
-    main()
+    sys.exit(register_subparser())
```
Comment (jimmylai) [inline fixit/cli/add_new_rule.py:135]: `register_subparser` always returns `None`.
To make it simpler, we can move sys.exit() inside `register_subparser`.

File: fixit/cli/add_new_rule.py [hunk 11]
```diff
@@ -86,32 +88,48 @@ def create_rule_file(file_path: Path, rule_name: str) -> None:
...
-    args = parser.parse_args()
 
+def register_subparser(parser: Optional[argparse._SubParsersAction] = None) -> None:
+    """Add parser or subparser for `add_new_rule` command."""
+    if parser is None:
+        add_rule_parser = argparse.ArgumentParser(
+            description="Creates a skeleton of adding new rule file",
+        )
+        _add_arguments(add_rule_parser)
+        _main(add_rule_parser.parse_args())
```
Comment (jimmylai) [inline fixit/cli/add_new_rule.py:114]:
```suggestion
        sys.exit(_main(add_rule_parser.parse_args()))
```

File: fixit/cli/main.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    # Register sub-commands
+    subparser = parser.add_subparsers(title="command")
+
+    add_new_rule.register_subparser(subparser)
+    apply_fix.register_subparser(subparser)
+    insert_suppressions.register_subparser(subparser)
+    run_rules.register_subparser(subparser)
+
+    args = parser.parse_args(argv)
+    args.subparser_fn(args)
```
Comment (jimmylai) [inline fixit/cli/main.py:40]:
```suggestion
    sys.exit(args.subparser_fn(args))
```

File: fixit/cli/main.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+
+if __name__ == "__main__":
+    sys.exit(main())
```
Comment (jimmylai) [inline fixit/cli/main.py:44]:
```suggestion
    main()
```

File: fixit/cli/add_new_rule.py (hunk 10 above)
Comment (acharles7) [reply fixit/cli/add_new_rule.py:135]: Isn't the `register_subparser()` always return `_main()` and `_main()` returns `int`

Comment (acharles7) [reply fixit/cli/add_new_rule.py:135]: But, your idea makes more sense. So, I'll go with your logic

**Pull request merged**